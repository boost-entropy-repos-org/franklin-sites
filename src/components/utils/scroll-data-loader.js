const scrollOffsetPercentage = 10;
const scrollOffsetFactor = 1 + scrollOffsetPercentage / 100;

export const isNotScrollable = (node) => {
  const { scrollHeight, offsetHeight } = node.current;
  return scrollHeight <= offsetHeight * scrollOffsetFactor;
};

export const isBottom = (node) => {
  const { scrollHeight, scrollTop, offsetHeight } = node.current;
  return scrollHeight - Math.ceil(scrollTop) <= offsetHeight * scrollOffsetFactor;
};

export const checkLoadMoreItems = (loading, hasMoreData, ref, setLoadMoreItems) => {
  if (!loading && hasMoreData && isBottom(ref)) {
    setLoadMoreItems(true);
  }
};

export const checkOnDataLoad = (
  hasMoreData,
  setLoading,
  onLoadMoreItems,
  setLoadMoreItems,
  ref,
) => {
  if (hasMoreData && isNotScrollable(ref)) {
    setLoading(true);
    onLoadMoreItems();
  } else {
    setLoading(false);
    setLoadMoreItems(false);
  }
};

export const checkOnLoadMoreItems = (loadMoreItems, setLoading, onLoadMoreItems) => {
  if (loadMoreItems) {
    setLoading(true);
    onLoadMoreItems();
  }
};
