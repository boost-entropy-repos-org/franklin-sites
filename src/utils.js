export function getLastIndexOfSubstringIgnoreCase(string, substring) {
  return string.toLowerCase().lastIndexOf(substring.toLowerCase());
}

export const getFlattenedPaths = (currentItems, value, path = []) => {
  let flattened = [];
  currentItems.forEach((node) => {
    const { items, ...thisNode } = node;
    const nodePath = [...path, thisNode];
    if (items) {
      const result = getFlattenedPaths(items, value, nodePath);
      if (result.length) {
        flattened = [...flattened, ...result];
      }
    } else if (!value || thisNode.value === value) {
      flattened = [...flattened, nodePath];
    }
  });
  return flattened;
};

function getSquashedPath(items, sep = ' / ') {
  return {
    value: items[items.length - 1].value,
    pathLabel: items.map(item => item.label).join(sep),
    itemLabel: items[items.length - 1].label,
    items,

  };
}

export function getSquashedArrayOfPaths(paths) {
  // console.log(paths);
  return paths.map(items => getSquashedPath(items));
}
