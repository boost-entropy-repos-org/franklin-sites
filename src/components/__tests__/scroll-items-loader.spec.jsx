import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import ScrollItemsLoader from '../scroll-items-loader';
import { fillArray } from '../../utils';

describe('ScrollItemsLoader', () => {
  const idKey = 'uuid';
  const items = fillArray(10, (element, index) => <p key={index}>{index}</p>);

  let onLoadMoreItems;
  beforeEach(() => {
    onLoadMoreItems = jest.fn();
  });

  test('should render ScrollItemsLoader', () => {
    const { asFragment } = render(
      <ScrollItemsLoader
        idKey={idKey}
        onLoadMoreItems={onLoadMoreItems}
        hasMoreItems
        items={items}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should request more data', () => {
    const { container } = render(
      <ScrollItemsLoader
        idKey={idKey}
        onLoadMoreItems={onLoadMoreItems}
        hasMoreItems
        items={items}
      />,
    );
    const scrollContainer = container.firstChild;
    fireEvent.scroll(scrollContainer, { target: { scrollY: 100 } });
    expect(onLoadMoreItems).toHaveBeenCalled();
  });

  test('should not request more data', () => {
    const { container } = render(
      <ScrollItemsLoader
        idKey={idKey}
        onLoadMoreItems={onLoadMoreItems}
        hasMoreItems={false}
        items={items}
      />,
    );
    const scrollContainer = container.firstChild;
    fireEvent.scroll(scrollContainer, { target: { scrollY: 100 } });
    expect(onLoadMoreItems).toHaveBeenCalledTimes(0);
  });
});
