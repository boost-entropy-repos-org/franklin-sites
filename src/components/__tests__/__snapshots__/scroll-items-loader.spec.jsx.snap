// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollItemsLoader should render ScrollItemsLoader 1`] = `
<DocumentFragment>
  <div
    class=" scroll-container"
  >
    <p>
      0
    </p>
    <p>
      1
    </p>
    <p>
      2
    </p>
    <p>
      3
    </p>
    <p>
      4
    </p>
    <p>
      5
    </p>
    <p>
      6
    </p>
    <p>
      7
    </p>
    <p>
      8
    </p>
    <p>
      9
    </p>
    <h4>
      Loading...
    </h4>
  </div>
</DocumentFragment>
`;
